src/System.Console/src/System/ConsolePal.Unix.cs

```diff
@@ -1141,15 +1228,103 @@ private static unsafe void Write(SafeFileHandle fd, byte* bufPtr, int count)
                         throw Interop.GetExceptionForIoErrno(errorInfo);
                     }
                 }
+                else
+                {
+                    if (!noCursorChange)
+                    {
+                        UpdatedCachedCursorPosition(bufPtr, bytesWritten, cursorVersion);
+                    }
+                }
 
                 count -= bytesWritten;
                 bufPtr += bytesWritten;
             }
         }
 
+        private static unsafe void UpdatedCachedCursorPosition(byte* bufPtr, int count, int cursorVersion)
+        {
+            lock (s_cursorGate)
+            {
+                if (cursorVersion != s_cursorVersion  ||  // the cursor was changed during the write by another operation
+                    !HasCachedCursorPosition()        ||  // we don't have a cursor position
+                    count > DefaultConsoleBufferSize)     // limit the amount of bytes we are willing to inspect
+                {
+                    InvalidateCachedCursorPosition();
+                    return;
+                }
+
+                int width = WindowWidth;
+                int height = WindowHeight;
+                int left = s_cursorLeft;
+                int top = s_cursorTop;
+
+                for (int i = 0; i < count; i++)
+                {
+                    byte c = bufPtr[i];
+                    if (c < 127 && c >= 32) // ASCII/UTF-8 characters that take up a single position
```

Comment:
This is essentially "visible ASCII characters", right?

Reply:
Yes.